src/Codec/Candid/Coerce.hs

```diff
@@ -5,253 +5,167 @@
-        c <- go t1 t2
-        return c
+    SubTypeM k1 k2 [Value]
 
 -- Look through refs
-go (RefT (Ref _ t1)) t2 = memo t1 t2
-go t1 (RefT (Ref _ t2)) = memo t1 t2
+go v (RefT (Ref _ t1)) t2 = go v t1 t2
+go v t1 (RefT (Ref _ t2)) = go v t1 t2
 
 -- Identity coercion for primitive values
-go NatT NatT = pure pure
-go Nat8T Nat8T = pure pure
-go Nat16T Nat16T = pure pure
-go Nat32T Nat32T = pure pure
-go Nat64T Nat64T = pure pure
-go IntT IntT = pure pure
-go Int8T Int8T = pure pure
-go Int16T Int16T = pure pure
-go Int32T Int32T = pure pure
-go Int64T Int64T = pure pure
-go Float32T Float32T = pure pure
-go Float64T Float64T = pure pure
-go BoolT BoolT = pure pure
-go TextT TextT = pure pure
-go NullT NullT = pure pure
-go PrincipalT PrincipalT = pure pure
+go v NatT NatT = pure v
+go v Nat8T Nat8T = pure v
+go v Nat16T Nat16T = pure v
+go v Nat32T Nat32T = pure v
+go v Nat64T Nat64T = pure v
+go v IntT IntT = pure v
+go v Int8T Int8T = pure v
+go v Int16T Int16T = pure v
+go v Int32T Int32T = pure v
+go v Int64T Int64T = pure v
+go v Float32T Float32T = pure v
+go v Float64T Float64T = pure v
+go v BoolT BoolT = pure v
+go v TextT TextT = pure v
+go v NullT NullT = pure v
+go v PrincipalT PrincipalT = pure v
 
 -- Nat <: Int
-go NatT IntT = pure $ \case
-    NatV n -> pure $ IntV (fromIntegral n)
-    v -> throwError $ show $ "Unexpected value" <+> pretty v <+> "while coercing nat <: int"
+go (NatV n) NatT IntT = pure $ IntV (fromIntegral n)
 
 -- t <: reserved
-go _ ReservedT = pure (const (pure ReservedV))
+go _ _ ReservedT = pure ReservedV
 
--- empty <: t
-go EmptyT _ = pure $ \v ->
-    throwError $ show $ "Unexpected value" <+> pretty v <+> "while coercing empty"
+-- empty <: t (actually just a special case of `v :/ t`)
+go v EmptyT _ = throwError $ show $ "Unexpected value" <+> pretty v <+> "while coercing empty"
 
 -- vec t1 <: vec t2
-go (VecT t1) (VecT t2) = do
-    c <- memo t1 t2
-    pure $ \case
-        VecV vs -> VecV <$> mapM c vs
-        v -> throwError $ show $ "Unexpected value" <+> pretty v <+> "while coercing vector"
+go (VecV vs) (VecT t1) (VecT t2) = VecV <$> mapM (\v -> go v t1 t2) vs
 
 -- Option: The normal rule
-go (OptT t1) (OptT t2) = lift (runExceptT (memo t1 t2)) >>= \case
-    Right c -> pure $ \case
-        OptV Nothing -> pure (OptV Nothing)
-        OptV (Just v) -> OptV . Just <$> c v
-        v -> throwError $ show $ "Unexpected value" <+> pretty v <+> "while coercing option"
-    Left _ -> pure (const (pure (OptV Nothing)))
+go (OptV Nothing)  (OptT _) (OptT _) = pure NullV
+go (OptV (Just v)) (OptT t1) (OptT t2) =
+    lift (runExceptT (go v t1 t2)) >>= \case
+        Right v' -> pure (OptV (Just v'))
+        Left _   -> pure (OptV Nothing)
 
 -- Option: The constituent rule
-go t (OptT t2) | not (isOptLike t2) = lift (runExceptT (memo t t2)) >>= \case
-    Right c -> pure $ \v -> OptV . Just <$> c v
-    Left _ -> pure (const (pure (OptV Nothing)))
+go v t1 (OptT t2) | not (isOptLike t2) =
```

Comment:
`isOptLike NullT` still returns `true` - is that correct?

Reply:
Yes, for the constituent it is, but not for the “is optional field/argument” rule, I think.

Reply:
Ok, but with https://github.com/dfinity/candid/pull/311 still having unresolved comments, I'm finding it hard to know what to implement ;->

Reply:
Ah, ok it think I get it now. And since my subtype check just allows `_ <: Opt T` (the admissable rule) with no further scrutiny, I guess I don't need to think about the constituent case specifically.

Reply:
Yes! See https://github.com/nomeata/haskell-candid/pull/20/files#diff-f852a1ad7d0ee80b5b01f6cda833856c03db823b49af0928c1030b92a5a5d283R113